### DISCUSSION TITLE
Need some advice on using Flutter Data for an offline-first app

### DISCUSSION THREAD

**Discussion by frellan** [flutterdata/flutter_data#261] [General]
Hey everyone 👋,

I'm diving into the `flutter_data` package for a project where I need to handle forms that are a bit on the longer side. The catch is, the internet isn't always reliable, and I'm trying to figure out the best way to keep everything synced up once the connection kicks back in. I've got a few questions on how `flutter_data` handles this stuff, especially when compared to something like [PowerSync](https://www.powersync.com/).

**Offline-first syncing:** How does `flutter_data` deal with offline-first data handling and syncing back to the server? My app's got these extensive forms, and I need to make sure they're saved locally then synced up once we're back online. Does `flutter_data` have a smooth and robust way to handle this? Sometimes the connection can come back up but be like 1 bar, then try and sync. Then fail, I then want it to retry next time and so on. This can occur multiple times during an entire day.

**Syncing whole JSON docs:** I'm thinking the simplest route might be just syncing up entire JSON docs to avoid any mess with model relationships, especially if things get updated while offline. Right? Or what would be a good approach here? I'm hoping to avoid any headaches with data conflicts or losses.

**Image syncing:** Another big part of my app involves syncing images. How's `flutter_data` in handling the syncing of images or large files once you get your connection back? Any built-in tricks or should I look into pairing it with something else for the heavy lifting here?

**flutter_data vs PowerSync:** I'm also curious how `flutter_data` stacks up against [PowerSync](https://www.powersync.com/) in terms of handling offline scenarios and syncing. I know PowerSync is a completely different kind of "product", I know, but just for comparison.

Would love to hear from anyone who's tackled similar challenges or has insights on making flutter_data work smoothly for a scenario like this. Any advice or pointers would be awesome.

Cheers!

**Comment by frank06** [comment]
I'm the author of FD and used it in a similar context in a complex app (offline first and sync). This is a complex issue and I don't think there's a smooth and easy way, but definitely possible. Have a look at the `OfflineOperation`  class
